### PR TITLE
[build] Making `config.h` fully available to CMake users

### PR DESCRIPTION
**Pull request by lzralbu**
Assuming config.h should be the main source of truth regarding raylib customization, I proposed the following changes:

* The CMakeOptions.txt file should expose all the flags found in config.h.
* The CompileDefinitions.cmake should treat the configuration values in config.h as the default ones.

To this end I created the ParseConfigHeader.cmake script to read config.h and generate lists which are in turn used by CMakeOptions.txt and CompileDefinitions.cmake to do their thing.

### CODE REVIEW

**Comment by raysan5** [comment]
@lzralbu It looks like a very big change on the CMake build system, not sure it could be potentially breaking for some users. 

I don't use or maintain CMake so someone with experience in the system should review this PR.

**Comment by lzralbu** [comment]
> @lzralbu is setting CUSTOMIZE_BUILD to On supposed to enable every single file format? because that's what your branch does for me...

No, regarding file formats it should only enable the SUPPORT_FILEFORMAT_* flags that are uncommented in the `config.h` file.

Could you share your cmake configuration step output?

Mine is like:

```
...
[cmake] -- SUPPORT_MODULE_RSHAPES=ON
[cmake] -- SUPPORT_MODULE_RTEXTURES=ON
[cmake] -- SUPPORT_MODULE_RTEXT=ON
[cmake] -- SUPPORT_MODULE_RMODELS=ON
[cmake] -- SUPPORT_CAMERA_SYSTEM=ON
[cmake] -- SUPPORT_GESTURES_SYSTEM=ON
[cmake] -- SUPPORT_RPRAND_GENERATOR=ON
[cmake] -- SUPPORT_MOUSE_GESTURES=ON
[cmake] -- SUPPORT_SSH_KEYBOARD_RPI=ON
[cmake] -- SUPPORT_WINMM_HIGHRES_TIMER=ON
[cmake] -- SUPPORT_PARTIALBUSY_WAIT_LOOP=ON
[cmake] -- SUPPORT_SCREEN_CAPTURE=ON
[cmake] -- SUPPORT_GIF_RECORDING=ON
[cmake] -- SUPPORT_COMPRESSION_API=ON
[cmake] -- SUPPORT_AUTOMATION_EVENTS=ON
[cmake] -- SUPPORT_FILEFORMAT_PNG=ON
[cmake] -- SUPPORT_IMAGE_EXPORT=ON
[cmake] -- SUPPORT_IMAGE_GENERATION=ON
[cmake] -- SUPPORT_IMAGE_MANIPULATION=ON
[cmake] -- SUPPORT_DEFAULT_FONT=ON
[cmake] -- SUPPORT_FILEFORMAT_TTF=ON
[cmake] -- SUPPORT_TEXT_MANIPULATION=ON
[cmake] -- SUPPORT_FONT_ATLAS_WHITE_REC=ON
[cmake] -- SUPPORT_FILEFORMAT_OBJ=ON
[cmake] -- SUPPORT_FILEFORMAT_GLTF=ON
[cmake] -- SUPPORT_MESH_GENERATION=ON
[cmake] -- SUPPORT_FILEFORMAT_WAV=ON
[cmake] -- SUPPORT_FILEFORMAT_OGG=ON
[cmake] -- SUPPORT_FILEFORMAT_MP3=ON
[cmake] -- SUPPORT_STANDARD_FILEIO=ON
[cmake] -- SUPPORT_TRACELOG=ON
[cmake] -- SUPPORT_TRACELOG_DEBUG=ON
...
```

There is no SUPPORT_FILEFORMAT_JPG=ON for example.

**Comment by Peter0x44** [comment]
@lzralbu I'm doing nothing special
```
[peter@peter build]$ cmake .. -DCUSTOMIZE_BUILD=On
-- Testing if -Werror=pointer-arith can be used -- compiles
-- Testing if -Werror=implicit-function-declaration can be used -- compiles
-- Testing if -fno-strict-aliasing can be used -- compiles
-- Using raylib's GLFW
-- Including Wayland support
-- Including X11 support
-- Building raylib static library
-- SUPPORT_MODULE_RSHAPES=ON
-- SUPPORT_MODULE_RTEXTURES=ON
-- SUPPORT_MODULE_RTEXT=ON
-- SUPPORT_MODULE_RMODELS=ON
-- SUPPORT_MODULE_RAUDIO=ON
-- SUPPORT_CAMERA_SYSTEM=ON
-- SUPPORT_GESTURES_SYSTEM=ON
-- SUPPORT_RPRAND_GENERATOR=ON
-- SUPPORT_MOUSE_GESTURES=ON
-- SUPPORT_SSH_KEYBOARD_RPI=ON
-- SUPPORT_WINMM_HIGHRES_TIMER=ON
-- SUPPORT_BUSY_WAIT_LOOP=ON
-- SUPPORT_PARTIALBUSY_WAIT_LOOP=ON
-- SUPPORT_SCREEN_CAPTURE=ON
-- SUPPORT_GIF_RECORDING=ON
-- SUPPORT_COMPRESSION_API=ON
-- SUPPORT_AUTOMATION_EVENTS=ON
-- SUPPORT_CUSTOM_FRAME_CONTROL=ON
-- SUPPORT_QUADS_DRAW_MODE=ON
-- SUPPORT_FILEFORMAT_PNG=ON
-- SUPPORT_FILEFORMAT_BMP=ON
-- SUPPORT_FILEFORMAT_TGA=ON
-- SUPPORT_FILEFORMAT_JPG=ON
-- SUPPORT_FILEFORMAT_GIF=ON
-- SUPPORT_FILEFORMAT_QOI=ON
-- SUPPORT_FILEFORMAT_PSD=ON
-- SUPPORT_FILEFORMAT_DDS=ON
-- SUPPORT_FILEFORMAT_HDR=ON
-- SUPPORT_FILEFORMAT_PIC=ON
-- SUPPORT_FILEFORMAT_KTX=ON
-- SUPPORT_FILEFORMAT_ASTC=ON
-- SUPPORT_FILEFORMAT_PKM=ON
-- SUPPORT_FILEFORMAT_PVR=ON
-- SUPPORT_FILEFORMAT_SVG=ON
-- SUPPORT_IMAGE_EXPORT=ON
-- SUPPORT_IMAGE_GENERATION=ON
-- SUPPORT_IMAGE_MANIPULATION=ON
-- SUPPORT_DEFAULT_FONT=ON
-- SUPPORT_FILEFORMAT_TTF=ON
-- SUPPORT_FILEFORMAT_FNT=ON
-- SUPPORT_FILEFORMAT_BDF=ON
-- SUPPORT_TEXT_MANIPULATION=ON
-- SUPPORT_FONT_ATLAS_WHITE_REC=ON
-- SUPPORT_FILEFORMAT_OBJ=ON
-- SUPPORT_FILEFORMAT_MTL=ON
-- SUPPORT_FILEFORMAT_IQM=ON
-- SUPPORT_FILEFORMAT_GLTF=ON
-- SUPPORT_FILEFORMAT_VOX=ON
-- SUPPORT_FILEFORMAT_M3D=ON
-- SUPPORT_MESH_GENERATION=ON
-- SUPPORT_FILEFORMAT_WAV=ON
-- SUPPORT_FILEFORMAT_OGG=ON
-- SUPPORT_FILEFORMAT_MP3=ON
-- SUPPORT_FILEFORMAT_QOA=ON
-- SUPPORT_FILEFORMAT_FLAC=ON
-- SUPPORT_FILEFORMAT_XM=ON
-- SUPPORT_FILEFORMAT_MOD=ON
-- SUPPORT_STANDARD_FILEIO=ON
-- SUPPORT_TRACELOG=ON
-- SUPPORT_TRACELOG_DEBUG=ON
-- Generated build type: Debug
-- Compiling with the flags:
--   PLATFORM=PLATFORM_DESKTOP
--   GRAPHICS=GRAPHICS_API_OPENGL_33
-- Building examples is enabled
-- Testing if -std=c11 can be used -- compiles
-- Configuring done (0.2s)
-- Generating done (0.3s)
-- Build files have been written to: /home/peter/bruh/raylib/build
```

**Comment by Peter0x44** [comment]
Honestly I think the way this is engineered is total overkill. New file formats don't get added often enough to make updating the default values in cmake a hassle, and nothing so terrible happens if one gets overlooked.

**Comment by lzralbu** [comment]
All this time I was testing my modifictions while using raylib as a subproject via add_subdirectory, but by cmake-configuring my branch as a top-level project gives me the same results as yours. I'll investigate further.

**Comment by lzralbu** [comment]
> @lzralbu I'm doing nothing special
> 
> ```
> [peter@peter build]$ cmake .. -DCUSTOMIZE_BUILD=On
> -- Testing if -Werror=pointer-arith can be used -- compiles
> -- Testing if -Werror=implicit-function-declaration can be used -- compiles
> -- Testing if -fno-strict-aliasing can be used -- compiles
> -- Using raylib's GLFW
> -- Including Wayland support
> -- Including X11 support
> -- Building raylib static library
> -- SUPPORT_MODULE_RSHAPES=ON
> -- SUPPORT_MODULE_RTEXTURES=ON
> -- SUPPORT_MODULE_RTEXT=ON
> -- SUPPORT_MODULE_RMODELS=ON
> -- SUPPORT_MODULE_RAUDIO=ON
> -- SUPPORT_CAMERA_SYSTEM=ON
> -- SUPPORT_GESTURES_SYSTEM=ON
> -- SUPPORT_RPRAND_GENERATOR=ON
> -- SUPPORT_MOUSE_GESTURES=ON
> -- SUPPORT_SSH_KEYBOARD_RPI=ON
> -- SUPPORT_WINMM_HIGHRES_TIMER=ON
> -- SUPPORT_BUSY_WAIT_LOOP=ON
> -- SUPPORT_PARTIALBUSY_WAIT_LOOP=ON
> -- SUPPORT_SCREEN_CAPTURE=ON
> -- SUPPORT_GIF_RECORDING=ON
> -- SUPPORT_COMPRESSION_API=ON
> -- SUPPORT_AUTOMATION_EVENTS=ON
> -- SUPPORT_CUSTOM_FRAME_CONTROL=ON
> -- SUPPORT_QUADS_DRAW_MODE=ON
> -- SUPPORT_FILEFORMAT_PNG=ON
> -- SUPPORT_FILEFORMAT_BMP=ON
> -- SUPPORT_FILEFORMAT_TGA=ON
> -- SUPPORT_FILEFORMAT_JPG=ON
> -- SUPPORT_FILEFORMAT_GIF=ON
> -- SUPPORT_FILEFORMAT_QOI=ON
> -- SUPPORT_FILEFORMAT_PSD=ON
> -- SUPPORT_FILEFORMAT_DDS=ON
> -- SUPPORT_FILEFORMAT_HDR=ON
> -- SUPPORT_FILEFORMAT_PIC=ON
> -- SUPPORT_FILEFORMAT_KTX=ON
> -- SUPPORT_FILEFORMAT_ASTC=ON
> -- SUPPORT_FILEFORMAT_PKM=ON
> -- SUPPORT_FILEFORMAT_PVR=ON
> -- SUPPORT_FILEFORMAT_SVG=ON
> -- SUPPORT_IMAGE_EXPORT=ON
> -- SUPPORT_IMAGE_GENERATION=ON
> -- SUPPORT_IMAGE_MANIPULATION=ON
> -- SUPPORT_DEFAULT_FONT=ON
> -- SUPPORT_FILEFORMAT_TTF=ON
> -- SUPPORT_FILEFORMAT_FNT=ON
> -- SUPPORT_FILEFORMAT_BDF=ON
> -- SUPPORT_TEXT_MANIPULATION=ON
> -- SUPPORT_FONT_ATLAS_WHITE_REC=ON
> -- SUPPORT_FILEFORMAT_OBJ=ON
> -- SUPPORT_FILEFORMAT_MTL=ON
> -- SUPPORT_FILEFORMAT_IQM=ON
> -- SUPPORT_FILEFORMAT_GLTF=ON
> -- SUPPORT_FILEFORMAT_VOX=ON
> -- SUPPORT_FILEFORMAT_M3D=ON
> -- SUPPORT_MESH_GENERATION=ON
> -- SUPPORT_FILEFORMAT_WAV=ON
> -- SUPPORT_FILEFORMAT_OGG=ON
> -- SUPPORT_FILEFORMAT_MP3=ON
> -- SUPPORT_FILEFORMAT_QOA=ON
> -- SUPPORT_FILEFORMAT_FLAC=ON
> -- SUPPORT_FILEFORMAT_XM=ON
> -- SUPPORT_FILEFORMAT_MOD=ON
> -- SUPPORT_STANDARD_FILEIO=ON
> -- SUPPORT_TRACELOG=ON
> -- SUPPORT_TRACELOG_DEBUG=ON
> -- Generated build type: Debug
> -- Compiling with the flags:
> --   PLATFORM=PLATFORM_DESKTOP
> --   GRAPHICS=GRAPHICS_API_OPENGL_33
> -- Building examples is enabled
> -- Testing if -std=c11 can be used -- compiles
> -- Configuring done (0.2s)
> -- Generating done (0.3s)
> -- Build files have been written to: /home/peter/bruh/raylib/build
> ```

Could you test again, please?

I guess the culprit was a regex that assumed inactive flags in config.h should start with exactly `// ` instead of `//` or `//                `, etc. My code formatter induced me to make such assumption.

**Comment by Peter0x44** [comment]
It seems to work correctly now, for that case. I checked it over again and I do have some questions/review comments, though.

File: cmake/LibraryConfigurations.cmake [hunk 1]
```diff
@@ -59,7 +59,7 @@ if (${PLATFORM} MATCHES "Desktop")
 elseif (${PLATFORM} MATCHES "Web")
     set(PLATFORM_CPP "PLATFORM_WEB")
     set(GRAPHICS "GRAPHICS_API_OPENGL_ES2")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -s USE_GLFW=3 -s ASSERTIONS=1 --profiling")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} --profiling")
```
Comment (Peter0x44) [inline cmake/LibraryConfigurations.cmake:62]: Why was this part for web changed?
It sounds outside of the scope of this PR, I'm not sure how it's related.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -42,21 +43,13 @@ if (NOT ${PLATFORM} MATCHES "Web")
...
 endif ()
 
+add_library(raylib ${raylib_sources} ${raylib_public_headers})
+
 # Sets additional platform options and link libraries for each platform
 # also selects the proper graphics API and version for that platform
 # Produces a variable LIBS_PRIVATE that will be used later
 include(LibraryConfigurations)
 
-if (USE_AUDIO)
```
Comment (Peter0x44) [inline src/CMakeLists.txt:50]: Why was this hunk deleted?
I see:
`USE_AUDIO is redundant in the presence of the already existent and more descriptive SUPPORT_MODULE_RAUDIO.`
Which is okay reasoning... but the option still shows up in ccmake or cmake-gui, and presumably does nothing.
It could maybe break existing users that rely on this (though I doubt there are many...)
I would be happy to review this in a separate PR, but I don't think it belongs in this one.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -69,7 +62,7 @@ else()
 endif()
 
 if (${PLATFORM} MATCHES "Web")
-    target_link_options(raylib PRIVATE "-sUSE_GLFW=3")
+    target_link_options(raylib INTERFACE "-sUSE_GLFW=3" "-sASSERTIONS=1")
```
Comment (Peter0x44) [inline src/CMakeLists.txt:65]: again, why was this changed in this PR? I think it should be submitted and reviewed separately.

**Review by Peter0x44 — CHANGES_REQUESTED** [review]

File: cmake/LibraryConfigurations.cmake (hunk 1 above)
Comment (lzralbu) [reply cmake/LibraryConfigurations.cmake:62]: Yeah, it does not belong here actually. I made a mistake when choosing branches.

Comment (Peter0x44) [reply cmake/LibraryConfigurations.cmake:62]: Okay. Please drop those changes from this PR.

File: src/CMakeLists.txt (hunk 3 above)
Comment (lzralbu) [reply src/CMakeLists.txt:65]: My bad. I guess you're right.

File: src/CMakeLists.txt (hunk 2 above)
Comment (Peter0x44) [reply src/CMakeLists.txt:50]: @lzralbu this would be more convenient to discuss elsewhere, like the raylib discord server. Ask there and I can help more directly.
I personally do this with `git rebase -i`.
You can change `pick` to `drop` in the editor for the commits you would like to remove.
After you've done this, you have to force push `git push --force`.
There's some more info available here:
https://git-rebase.io

**Comment by Peter0x44** [comment]
Works properly for me now. Approved.

**Comment by raysan5** [comment]
@lzralbu Thank you very much for the improvement! Despite the added build complexity, it improves alignment with raylib `config.h`, there were some issues in the past because of that need to update two files separately. Thanks!

@Peter0x44 Thank you very much for the review and testing of this BIG redesign!
